### PR TITLE
Automate Protected Devices to run `particle usb` commands

### PR DESCRIPTION
**Pull request by keeramis**
## Description

Support Protected Devices while running `particle usb xxx` commands
All of the `particle usb xxx` commands enable Protection on the device after the command is finished except these two:
- `particle usb reset`
- `particle usb dfu`

Also, added a fix for `particle usb cloud-status device-name` which was throwing an error due to an already existing bug when using `device-name`

## How to Test

Run `particle usb xxx` commands on Protected, Protected (Service Mode), and Open Devices.
Run them when the device is in normal mode and dfu mode. 
- Are you seeing the expected output?
- Are you seeing the device recover if it errors?
- Does it show helpful error messages?

## Related Issues / Discussions

https://github.com/particle-iot/particle-cli/pull/754
https://app.shortcut.com/particle/story/129067/add-generic-helpers-for-a-device-to-put-into-service-mode-and-back-to-perform-a-cli-operation

## Completeness

- [x] User is totes amazing for contributing!
- [ ] Contributor has signed [CLA](https://docs.google.com/a/particle.io/forms/d/1_2P-vRKGUFg5bmpcKLHO_qNZWGi5HKYnfrrkd-sbZoA/viewform)
- [ ] Problem and solution clearly stated
- [ ] Tests have been provided
- [ ] Docs have been updated
- [ ] CI is passing

### CODE REVIEW

File: src/cmd/usb-util.js [hunk 1]
```diff
@@ -124,8 +124,11 @@ async function executeWithUsbDevice({ args, func, dfuMode = false } = {}) {
 	} finally {
 		if (deviceIsProtected) {
 			device = await reopenDevice(device);
-			// FIXME: ignore errors
-			await deviceProtectionHelper.turnOffServiceMode(device);
+			try {
+				await deviceProtectionHelper.turnOffServiceMode(device);
+			} catch (error) {
+				// FIXME: ignore errors
```
Comment (monkbroc) [inline src/cmd/usb-util.js:130]: You can get rid of the FIXME.
Which use case did you see that needs errors here to be ignored?

File: src/cmd/usb-util.js [hunk 2]
```diff
@@ -124,8 +124,11 @@ async function executeWithUsbDevice({ args, func, dfuMode = false } = {}) {
...
+			}
 		}
```
Comment (keeramis) [inline src/cmd/usb-util.js:132]: `executeWithUsbDevice` is called from `forEachUsbDevice`, and this error will be propagated to `forEachUsbDevice` and if not handled properly, it will show an internal error to the user (such as `IN control transfer failed` which is likely useless to the user because they cant act on it)

File: src/cmd/usb.js [hunk 3]
```diff
@@ -186,52 +188,56 @@ module.exports = class UsbCommand extends CLICommandBase {
...
+			if (Date.now() - started >= timeout) {
+				throw new Error('timed-out waiting for status...');
+			}
+		} else {
+			try {
+				status = await device.getCloudConnectionStatus();
+			} catch (error) {
+				// FIXME
+				throw new Error(error);
+			}
```
Comment (monkbroc) [inline src/cmd/usb.js:237]: This is unnecessary. If you don't want to catch the error, remove the try and catch blocks.

File: test/e2e/usb.e2e.js [hunk 4]
```diff
@@ -179,16 +181,21 @@ describe('USB Commands [@device]', function cliUSBCommands(){
 			expect(exitCode).to.equal(0);
 		});
 
-		it('Lists connected devices when signed-in to a foreign account', async () => {
+		xit('Lists connected devices when signed-in to a foreign account', async () => {
+			await cli.logout();
 			await cli.loginToForeignAcct();
 			const { stdout, stderr, exitCode } = await cli.run(args);
+			console.log('stdout:', stdout);
```
Comment (monkbroc) [inline test/e2e/usb.e2e.js:188]: Remember to remove console logs

File: test/e2e/usb.e2e.js [hunk 5]
```diff
@@ -179,16 +181,21 @@ describe('USB Commands [@device]', function cliUSBCommands(){
 			expect(exitCode).to.equal(0);
 		});
 
-		it('Lists connected devices when signed-in to a foreign account', async () => {
+		xit('Lists connected devices when signed-in to a foreign account', async () => {
```
Comment (monkbroc) [inline test/e2e/usb.e2e.js:184]: Why does this test not work anymore?

File: test/lib/cli.js [hunk 6]
```diff
@@ -182,7 +184,7 @@ module.exports.getCloudConnectionStatus = async () => {
 
 module.exports.waitUntilOnline = async () => {
 	const { run } = module.exports;
-	const args = ['usb', 'cloud-status', DEVICE_ID, '--until', 'connected', '--timeout', 5 * 60 * 1000];
+	const args = ['usb', 'cloud-status', DEVICE_ID, '--until', 'connected', '--timeout', 5000];
```
Comment (monkbroc) [inline test/lib/cli.js:187]: Waiting 5 seconds for the device to connect to the cloud during E2E tests isn't enough. Why the change?

**Review by monkbroc — COMMENTED** [review]
Looks good but I have some questions on the changes to tests

File: src/cmd/usb-util.js (hunk 1 above)
Comment (keeramis) [reply src/cmd/usb-util.js:130]: @monkbroc 
`executeWithUsbDevice` is called from `forEachUsbDevice`, and this error will be propagated to `forEachUsbDevice` and if not handled properly, it will show an internal error to the user (such as IN control transfer failed which is likely useless to the user because they cant act on it)

Use case: Enabling Protection on the device immediately after running the safe mode command caused an `IN control transfer failed` error. I added a `waitForDeviceToReboot` after with safe mode command to let it work correctly but I realized that an error (if any) can propagate to `forEachUsbDevice` when using `turnOffServiceMode`

Comment (keeramis) [reply src/cmd/usb-util.js:130]: Ignoring this error means that an internal error will not be thrown to the user, and device will likely be in Service Mode if any error occurs here that is preventing turning off Service Mode

File: src/cmd/usb.js [hunk 7]
```diff
@@ -186,52 +187,51 @@ module.exports = class UsbCommand extends CLICommandBase {
...
+		try {
+			status = await executeWithUsbDevice({
+				args: { idOrName: device }, // device here is the id
+				func: (dev) => this._cloudStatus(dev, until, timeout),
+			});
+		} catch (error) {
+			this.stopSpin();
+			throw new Error(error.message);
+		} finally {
+			this.stopSpin();
```
Comment (hugomontero) [inline src/cmd/usb.js:204]: `finally` will always be executed, so you can remove `this.stopSpin()` from catch since you are doing it in the `finally` block

Check here 👉  [try-catch](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/try...catch)

File: src/cmd/usb.js [hunk 8]
```diff
@@ -260,7 +260,7 @@ module.exports = class UsbCommand extends CLICommandBase {
 		// define output array with logs to prevent interleaving with the spinner
 		let output = [];
 
-		await forEachUsbDevice(args, usbDevice => {
+		await forEachUsbDevice(args, async (usbDevice) => {
```
Comment (hugomontero) [inline src/cmd/usb.js:263]: The structure here didn't change, why we put the `async` here?

File: src/cmd/usb.js [hunk 9]
```diff
@@ -186,52 +187,51 @@ module.exports = class UsbCommand extends CLICommandBase {
...
-			.catch(error => {
-				if (until){
-					return deviceMgr.close()
-						.then(() => delay(1000))
-						.then(() => this.cloudStatus(args, started));
-				}
-				throw error;
-			})
-			.finally(() => deviceMgr.close());
+		return status;
```
Comment (hugomontero) [inline src/cmd/usb.js:234]: How about:

```
const started = Date.now();

if (until) {
	while (Date.now() - started < timeout) {
		try {
			let status = await device.getCloudConnectionStatus();
			status = status.toLowerCase();
			if (status === until) {
				return status;
			}
		} catch (error) {
			// ignore error and keep trying until timeout has elapsed
		}
	}
	throw new Error('timed-out waiting for status...');
} else {
	return device.getCloudConnectionStatus();
}
```

Comment (keeramis) [reply src/cmd/usb.js:244]: this works too. I preferred a single point of return

File: src/cmd/usb.js (hunk 7 above)
Comment (keeramis) [reply src/cmd/usb.js:214]: I wanted to ensure the error thrown does not interfere with the spinner

File: src/cmd/usb-util.js [hunk 10]
```diff
@@ -432,10 +467,15 @@ async function forEachUsbDevice(args, func, { dfuMode = false } = {}){
...
-							args: { idOrName : usbDevice.id },
+							args: { idOrName : usbDevice.id, api: args.api, auth: args.auth },
 							func,
 							dfuMode
 						});
+						// This is particularly useful if the device handle is lost in the executed function
+						// Example: particle usb reset
+						if (usbDevice?.isOpen) {
+							await usbDevice.close();
+						}
```
Comment (keeramis) [inline src/cmd/usb-util.js:478]: Still unclear to me why the handle is lost in the executeWithUsbDevice but this was needed to close the hanging handle

File: test/e2e/usb-protected.e2e.js [hunk 11]
```diff
@@ -0,0 +1,440 @@
+const capitalize = require('lodash/capitalize');
```
Comment (keeramis) [inline test/e2e/usb-protected.e2e.js:1]: Comment for this file as a whole. I intentionally included a few redundant checks in the `after` of the previous describe block as well as `before` of the current describe block for increases robustness

File: src/cmd/usb-util.js [hunk 12]
```diff
@@ -123,9 +124,16 @@ async function executeWithUsbDevice({ args, func, dfuMode = false } = {}) {
...
+			device = await reopenDevice({ id: deviceId });
+			if (!device.isInDfuMode) {
+				device = await waitForDeviceToReboot(device.id);
+			}
+			try {
+				// Only works for 6.1.2 and later
+				await deviceProtectionHelper.turnOffServiceMode(device);
+			} catch (error) {
+				// FIXME: ignore errors
+			}
```
Comment (keeramis) [inline src/cmd/usb-util.js:136]: I added this to ensure the device is ready for further operations after it finished the respective CLI operation. I forgot which use case needed this change but it was during e2e tests that I found out that this is a good check to have. IMO it adds more robustness. Please let me know if anyone disagrees.
```
if (!device.isInDfuMode) {
        device = await waitForDeviceToReboot(device.id);
}
```

File: src/cmd/usb-util.js [hunk 13]
```diff
@@ -134,6 +137,33 @@ async function executeWithUsbDevice({ args, func, dfuMode = false } = {}) {
...
+			// Check device readiness
+			await device.getDeviceId();
+			return device;
+		} catch (error) {
+			// ignore error
+		}
+	}
+}
+
+async function _delay(ms){
```
Comment (hugomontero) [inline src/cmd/usb-util.js:168]: We already have a delay function 👉 https://github.com/particle-iot/particle-cli/blob/0fcc1d661ad7331d20eb9578a798a20cfcfe2f3b/src/lib/utilities.js#L139 Can you use it instead of creating a new one?

File: test/lib/cli.js [hunk 14]
```diff
@@ -52,9 +52,11 @@ module.exports.login = () => {
 	return run(['login', '-u', USERNAME, '-p', PASSWORD], { reject: true });
 };
 
-module.exports.loginToForeignAcct = () => {
+module.exports.loginToForeignAcct = async () => {
 	const { run } = module.exports;
-	return run(['login', '-u', FOREIGN_USERNAME, '-p', FOREIGN_PASSWORD], { reject: true });
+	await run(['login', '-u', FOREIGN_USERNAME, '-p', FOREIGN_PASSWORD], { reject: true });
+	const s = await run(['whoami'], { reject: true });
+	console.log('Logged in as:', s);
```
Comment (hugomontero) [inline test/lib/cli.js]: Are those lines required for this test?

File: src/cmd/usb.js (hunk 7 above)
Comment (hugomontero) [reply src/cmd/usb.js]: It shouldn't since `finally` will always be executed. 
image

**Review by hugomontero — APPROVED** [review]
Just few comments but looks good!

File: src/cmd/usb-util.js (hunk 10 above)
Comment (monkbroc) [reply src/cmd/usb-util.js:478]: There's a `device.close()` at end of `executeWithUsbDevice`. This seems extra

**Review by monkbroc — APPROVED** [review]
Looks good! I tested a couple commands and it worked well.
